File: interop/javascript/test_reflection.js
#!/usr/bin/env node
/**
 * Comprehensive test suite for Flow JavaScript bindings reflection API
 */

const flow = require('./index.js');
const assert = require('assert');

let testsPassed = 0;
let testsFailed = 0;

function test(name, fn) {
    try {
        fn();
        console.log(`✓ ${name}`);
        testsPassed++;
    } catch (error) {
        console.log(`✗ ${name}`);
        console.log(`  ${error.message}`);
        testsFailed++;
    }
}

function assertEqual(actual, expected, message) {
    if (actual !== expected) {
        throw new Error(`${message}: expected ${expected}, got ${actual}`);
    }
}

function assertArrayEqual(actual, expected, message) {
    if (JSON.stringify(actual.sort()) !== JSON.stringify(expected.sort())) {
        throw new Error(`${message}: expected ${JSON.stringify(expected)}, got ${JSON.stringify(actual)}`);
    }
}

function assertTrue(condition, message) {
    if (!condition) {
        throw new Error(message);
    }
}

function assertContains(haystack, needle, message) {
    if (!haystack.includes(needle)) {
        throw new Error(`${message}: '${haystack}' does not contain '${needle}'`);
    }
}

console.log('============================================================');
console.log('Flow JavaScript Reflection API Tests');
console.log('============================================================\n');

// Test 1: Function count
test('test_function_count', () => {
    const code = `
func add(a: int, b: int) -> int {
    return a + b;
}

func subtract(x: int, y: int) -> int {
    return x - y;
}

func multiply(m: int, n: int) -> int {
    return m * n;
}
`;
    
    const module = flow.compile(code);
    const count = module.functionCount();
    
    assertEqual(count, 3, 'Should have 3 functions');
});

// Test 2: List functions
test('test_list_functions', () => {
    const code = `
func greet(name: string) -> string {
    return "Hello, " + name;
}

func square(x: int) -> int {
    return x * x;
}

func is_positive(n: int) -> bool {
    return n > 0;
}
`;
    
    const module = flow.compile(code);
    const functions = module.listFunctions();
    
    assertEqual(functions.length, 3, 'Should have 3 function names');
    assertArrayEqual(functions, ['greet', 'square', 'is_positive'], 'Function names should match');
});

// Test 3: Function info with parameters
test('test_function_info_with_parameters', () => {
    const code = `
func add(a: int, b: int) -> int {
    return a + b;
}
`;
    
    const module = flow.compile(code);
    const info = module.functionInfo('add');
    
    assertEqual(info.name, 'add', 'Function name should be add');
    assertEqual(info.return_type, 'int', 'Return type should be int');
    assertEqual(info.parameters.length, 2, 'Should have 2 parameters');
    assertEqual(info.parameters[0].name, 'a', 'First param name');
    assertEqual(info.parameters[0].type, 'int', 'First param type');
    assertEqual(info.parameters[1].name, 'b', 'Second param name');
    assertEqual(info.parameters[1].type, 'int', 'Second param type');
});

// Test 4: Function info with string parameter
test('test_function_info_string_parameter', () => {
    const code = `
func greet(name: string) -> string {
    return "Hello, " + name;
}
`;
    
    const module = flow.compile(code);
    const info = module.functionInfo('greet');
    
    assertEqual(info.name, 'greet', 'Function name');
    assertEqual(info.return_type, 'string', 'Return type');
    assertEqual(info.parameters.length, 1, 'Parameter count');
    assertEqual(info.parameters[0].name, 'name', 'Param name');
    assertEqual(info.parameters[0].type, 'string', 'Param type');
});

// Test 5: Function info with no parameters
test('test_function_info_no_parameters', () => {
    const code = `
func get_pi() -> float {
    return 3.14159;
}
`;
    
    const module = flow.compile(code);
    const info = module.functionInfo('get_pi');
    
    assertEqual(info.name, 'get_pi', 'Function name');
    assertEqual(info.return_type, 'float', 'Return type');
    assertEqual(info.parameters.length, 0, 'Should have no parameters');
});

// Test 6: Function info nonexistent
test('test_function_info_nonexistent', () => {
    const code = `
func test() -> int {
    return 42;
}
`;
    
    const module = flow.compile(code);
    
    let threw = false;
    try {
        module.functionInfo('nonexistent');
    } catch (e) {
        threw = true;
        assertContains(e.message, 'not found', 'Error message should mention not found');
    }
    
    assertTrue(threw, 'Should throw error for non-existent function');
});

// Test 7: Inspect functions
test('test_inspect_functions', () => {
    const code = `
func add(a: int, b: int) -> int {
    return a + b;
}

func greet(name: string) -> string {
    return "Hello";
}
`;
    
    const module = flow.compile(code);
    const inspection = module.inspect();
    
    assertContains(inspection, 'add(a: int, b: int) -> int', 'Should contain add signature');
    assertContains(inspection, 'greet(name: string) -> string', 'Should contain greet signature');
    assertContains(inspection, '2 function(s)', 'Should mention 2 functions');
});

// Test 8: Boolean return types
test('test_boolean_return_types', () => {
    const code = `
func is_positive(x: int) -> bool {
    return x > 0;
}

func is_even(n: int) -> bool {
    return n % 2 == 0;
}
`;
    
    const module = flow.compile(code);
    const info = module.functionInfo('is_positive');
    
    assertEqual(info.return_type, 'bool', 'Return type should be bool');
    assertEqual(module.functionCount(), 2, 'Should have 2 functions');
});

// Test 9: Large module
test('test_large_module', () => {
    let code = '';
    for (let i = 0; i < 20; i++) {
        code += `func func${i}(x: int) -> int { return x * ${i}; }\n`;
    }
    
    const module = flow.compile(code);
    const count = module.functionCount();
    
    assertEqual(count, 20, 'Should have 20 functions');
    
    const functions = module.listFunctions();
    assertEqual(functions.length, 20, 'Should list 20 functions');
    
    const info0 = module.functionInfo('func0');
    assertEqual(info0.name, 'func0', 'First function name');
    
    const info19 = module.functionInfo('func19');
    assertEqual(info19.name, 'func19', 'Last function name');
});

// Test 10: Functions still callable after reflection
test('test_functions_still_callable_after_reflection', () => {
    const code = `
func add(a: int, b: int) -> int {
    return a + b;
}

func multiply(x: int, y: int) -> int {
    return x * y;
}
`;
    
    const module = flow.compile(code);
    
    // Do reflection
    const count = module.functionCount();
    assertEqual(count, 2, 'Should have 2 functions');
    
    const info = module.functionInfo('add');
    assertEqual(info.name, 'add', 'Function name should be add');
    
    // Now call the functions to ensure reflection didn't break them
    const result1 = module.call('add', 10, 20);
    assertEqual(result1, 30, 'add(10, 20) should return 30');
    
    const result2 = module.call('multiply', 7, 6);
    assertEqual(result2, 42, 'multiply(7, 6) should return 42');
});

// Test 11: Mixed parameter types
test('test_mixed_parameter_types', () => {
    const code = `
func process(a: int, b: float, c: string, d: bool) -> string {
    return "processed";
}
`;
    
    const module = flow.compile(code);
    const info = module.functionInfo('process');
    
    assertEqual(info.parameters.length, 4, 'Should have 4 parameters');
    assertEqual(info.parameters[0].type, 'int', 'First param type');
    assertEqual(info.parameters[1].type, 'float', 'Second param type');
    assertEqual(info.parameters[2].type, 'string', 'Third param type');
    assertEqual(info.parameters[3].type, 'bool', 'Fourth param type');
});

// Test 12: Empty module
test('test_empty_module', () => {
    const code = '// Just a comment';
    
    const module = flow.compile(code);
    const count = module.functionCount();
    
    assertEqual(count, 0, 'Empty module should have 0 functions');
    
    const functions = module.listFunctions();
    assertEqual(functions.length, 0, 'Empty module should list 0 functions');
    
    const inspection = module.inspect();
    assertContains(inspection.toLowerCase(), 'no functions', 'Should mention no functions');
});

// Test 13: Float return type
test('test_float_return_type', () => {
    const code = `
func calculate_pi() -> float {
    return 3.14159;
}
`;
    
    const module = flow.compile(code);
    const info = module.functionInfo('calculate_pi');
    
    assertEqual(info.return_type, 'float', 'Return type should be float');
    assertEqual(info.parameters.length, 0, 'Should have no parameters');
});

// Test 14: Multiple string functions
test('test_multiple_string_functions', () => {
    const code = `
func concat(a: string, b: string) -> string {
    return a + b;
}

func repeat(s: string, n: int) -> string {
    return s;
}
`;
    
    const module = flow.compile(code);
    
    const info1 = module.functionInfo('concat');
    assertEqual(info1.parameters.length, 2, 'concat should have 2 parameters');
    assertEqual(info1.parameters[0].type, 'string', 'First param type');
    assertEqual(info1.parameters[1].type, 'string', 'Second param type');
    
    const info2 = module.functionInfo('repeat');
    assertEqual(info2.parameters.length, 2, 'repeat should have 2 parameters');
    assertEqual(info2.parameters[0].type, 'string', 'First param type');
    assertEqual(info2.parameters[1].type, 'int', 'Second param type');
});

// Print summary
console.log('\n============================================================');
console.log(`Tests passed: ${testsPassed}`);
console.log(`Tests failed: ${testsFailed}`);
console.log('============================================================');

process.exit(testsFailed > 0 ? 1 : 0);

File: interop/ruby/test_reflection.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Comprehensive test suite for Flow Ruby bindings reflection API

require_relative 'lib/flow'
require 'minitest/autorun'

class FlowReflectionTest < Minitest::Test
  def setup
    @runtime = Flow::Runtime.new
  end

  def test_function_count
    code = <<~FLOW
      func add(a: int, b: int) -> int {
        return a + b;
      }
      
      func subtract(x: int, y: int) -> int {
        return x - y;
      }
      
      func multiply(m: int, n: int) -> int {
        return m * n;
      }
    FLOW
    
    mod = Flow::Module.compile(@runtime, code)
    assert_equal 3, mod.function_count, "Should have 3 functions"
  end

  def test_list_functions
    code = <<~FLOW
      func greet(name: string) -> string {
        return "Hello, " + name;
      }
      
      func square(x: int) -> int {
        return x * x;
      }
      
      func is_positive(n: int) -> bool {
        return n > 0;
      }
    FLOW
    
    mod = Flow::Module.compile(@runtime, code)
    functions = mod.list_functions
    
    assert_equal 3, functions.size, "Should have 3 function names"
    assert_includes functions, 'greet', "Should include 'greet'"
    assert_includes functions, 'square', "Should include 'square'"
    assert_includes functions, 'is_positive', "Should include 'is_positive'"
  end

  def test_function_info_with_parameters
    code = <<~FLOW
      func add(a: int, b: int) -> int {
        return a + b;
      }
    FLOW
    
    mod = Flow::Module.compile(@runtime, code)
    info = mod.function_info('add')
    
    assert_equal 'add', info[:name], "Function name should be 'add'"
    assert_equal 'int', info[:return_type], "Return type should be 'int'"
    assert_equal 2, info[:parameters].size, "Should have 2 parameters"
    
    assert_equal 'a', info[:parameters][0][:name], "First param name"
    assert_equal 'int', info[:parameters][0][:type], "First param type"
    assert_equal 'b', info[:parameters][1][:name], "Second param name"
    assert_equal 'int', info[:parameters][1][:type], "Second param type"
  end

  def test_function_info_string_parameter
    code = <<~FLOW
      func greet(name: string) -> string {
        return "Hello, " + name;
      }
    FLOW
    
    mod = Flow::Module.compile(@runtime, code)
    info = mod.function_info('greet')
    
    assert_equal 'greet', info[:name]
    assert_equal 'string', info[:return_type]
    assert_equal 1, info[:parameters].size
    assert_equal 'name', info[:parameters][0][:name]
    assert_equal 'string', info[:parameters][0][:type]
  end

  def test_function_info_no_parameters
    code = <<~FLOW
      func get_pi() -> float {
        return 3.14159;
      }
    FLOW
    
    mod = Flow::Module.compile(@runtime, code)
    info = mod.function_info('get_pi')
    
    assert_equal 'get_pi', info[:name]
    assert_equal 'float', info[:return_type]
    assert_equal 0, info[:parameters].size, "Should have no parameters"
  end

  def test_function_info_nonexistent
    code = <<~FLOW
      func test() -> int {
        return 42;
      }
    FLOW
    
    mod = Flow::Module.compile(@runtime, code)
    
    assert_raises Flow::FlowError do
      mod.function_info('nonexistent')
    end
  end

  def test_inspect_functions
    code = <<~FLOW
      func add(a: int, b: int) -> int {
        return a + b;
      }
      
      func greet(name: string) -> string {
        return "Hello";
      }
    FLOW
    
    mod = Flow::Module.compile(@runtime, code)
    inspection = mod.inspect_functions
    
    assert_match(/add\(a: int, b: int\) -> int/, inspection, "Should contain 'add' signature")
    assert_match(/greet\(name: string\) -> string/, inspection, "Should contain 'greet' signature")
    assert_match(/2 function\(s\)/, inspection, "Should mention 2 functions")
  end

  def test_boolean_return_types
    code = <<~FLOW
      func is_positive(x: int) -> bool {
        return x > 0;
      }
      
      func is_even(n: int) -> bool {
        return n % 2 == 0;
      }
    FLOW
    
    mod = Flow::Module.compile(@runtime, code)
    
    info = mod.function_info('is_positive')
    assert_equal 'bool', info[:return_type]
    
    functions = mod.list_functions
    assert_equal 2, functions.size
  end

  def test_large_module
    # Generate a module with many functions
    code = (0...20).map do |i|
      "func func#{i}(x: int) -> int {\n  return x * #{i};\n}\n"
    end.join("\n")
    
    mod = Flow::Module.compile(@runtime, code)
    
    count = mod.function_count
    assert_equal 20, count, "Should have 20 functions"
    
    functions = mod.list_functions
    assert_equal 20, functions.size, "Should list 20 functions"
    
    # Check specific functions
    info0 = mod.function_info('func0')
    assert_equal 'func0', info0[:name]
    
    info19 = mod.function_info('func19')
    assert_equal 'func19', info19[:name]
  end

  def test_functions_still_callable_after_reflection
    code = <<~FLOW
      func add(a: int, b: int) -> int {
        return a + b;
      }
      
      func multiply(x: int, y: int) -> int {
        return x * y;
      }
    FLOW
    
    mod = Flow::Module.compile(@runtime, code)
    
    # Do reflection
    count = mod.function_count
    assert_equal 2, count
    
    info = mod.function_info('add')
    assert_equal 'add', info[:name]
    
    # Now call the functions to ensure reflection didn't break them
    result = mod.call('add', 10, 20)
    assert_equal 30, result, "add(10, 20) should return 30"
    
    result = mod.call('multiply', 7, 6)
    assert_equal 42, result, "multiply(7, 6) should return 42"
  end

  def test_mixed_parameter_types
    code = <<~FLOW
      func process(a: int, b: float, c: string, d: bool) -> string {
        return "processed";
      }
    FLOW
    
    mod = Flow::Module.compile(@runtime, code)
    info = mod.function_info('process')
    
    assert_equal 4, info[:parameters].size
    assert_equal 'int', info[:parameters][0][:type]
    assert_equal 'float', info[:parameters][1][:type]
    assert_equal 'string', info[:parameters][2][:type]
    assert_equal 'bool', info[:parameters][3][:type]
  end

  def test_empty_module
    code = "// Just a comment"
    
    mod = Flow::Module.compile(@runtime, code)
    
    count = mod.function_count
    assert_equal 0, count, "Empty module should have 0 functions"
    
    functions = mod.list_functions
    assert_equal 0, functions.size, "Empty module should list 0 functions"
    
    inspection = mod.inspect_functions
    assert_match(/no functions/i, inspection, "Should mention no functions")
  end

  def test_file_loaded_module
    # Create a temporary Flow file
    require 'tempfile'
    
    file = Tempfile.new(['test_reflect', '.flow'])
    begin
      file.write(<<~FLOW)
        func square(x: int) -> int {
          return x * x;
        }
        
        func cube(x: int) -> int {
          return x * x * x;
        }
      FLOW
      file.close
      
      mod = Flow::Module.load(@runtime, file.path)
      functions = mod.list_functions
      
      assert_includes functions, 'square', "'square' should be in loaded module"
      assert_includes functions, 'cube', "'cube' should be in loaded module"
      
      info = mod.function_info('square')
      assert_equal 'square', info[:name]
    ensure
      file.unlink
    end
  end

  def test_float_return_type
    code = <<~FLOW
      func calculate_pi() -> float {
        return 3.14159;
      }
    FLOW
    
    mod = Flow::Module.compile(@runtime, code)
    info = mod.function_info('calculate_pi')
    
    assert_equal 'float', info[:return_type]
    assert_equal 0, info[:parameters].size
  end

  def test_multiple_string_functions
    code = <<~FLOW
      func concat(a: string, b: string) -> string {
        return a + b;
      }
      
      func repeat(s: string, n: int) -> string {
        return s;
      }
    FLOW
    
    mod = Flow::Module.compile(@runtime, code)
    
    info1 = mod.function_info('concat')
    assert_equal 2, info1[:parameters].size
    assert_equal 'string', info1[:parameters][0][:type]
    assert_equal 'string', info1[:parameters][1][:type]
    
    info2 = mod.function_info('repeat')
    assert_equal 2, info2[:parameters].size
    assert_equal 'string', info2[:parameters][0][:type]
    assert_equal 'int', info2[:parameters][1][:type]
  end
end

# Run the tests if this file is executed directly
if __FILE__ == $0
  puts "\n" + "=" * 60
  puts "Flow Ruby Reflection API Tests"
  puts "=" * 60 + "\n"
end

